Modules/_testexternalinspection.c

```diff
@@ -433,44 +469,623 @@ read_string(pid_t pid, _Py_DebugOffsets* debug_offsets, void* address, char* buf
     return 0;
 }
 
-void*
-get_py_runtime(pid_t pid)
+
+static inline int
+read_ptr(pid_t pid, uintptr_t address, uintptr_t *ptr_addr)
 {
-#if defined(__linux__)
-    return get_py_runtime_linux(pid);
-#elif defined(__APPLE__) && TARGET_OS_OSX
-    return get_py_runtime_macos(pid);
-#else
-    return NULL;
-#endif
+    int bytes_read = read_memory(pid, address, sizeof(void*), ptr_addr);
+    if (bytes_read == -1) {
+        return -1;
+    }
+    return 0;
+}
+
+static inline int
+read_ssize_t(pid_t pid, uintptr_t address, Py_ssize_t *size)
+{
+    int bytes_read = read_memory(pid, address, sizeof(Py_ssize_t), size);
+    if (bytes_read == -1) {
+        return -1;
+    }
+    return 0;
 }
 
 static int
-parse_code_object(
-        int pid,
-        PyObject* result,
-        struct _Py_DebugOffsets* offsets,
-        void* address,
-        void** previous_frame)
+read_py_ptr(pid_t pid, uintptr_t address, uintptr_t *ptr_addr)
 {
-    void* address_of_function_name;
-    read_memory(
+    if (read_ptr(pid, address, ptr_addr)) {
+        return -1;
+    }
+    *ptr_addr &= ~Py_TAG_BITS;
+    return 0;
+}
+
+static int
+read_char(pid_t pid, uintptr_t address, char *result)
+{
+    int bytes_read = read_memory(pid, address, sizeof(char), result);
+    if (bytes_read < 0) {
```

Comment:
Any reason why this one uses `< 0` while `read_ssize_t` uses `== -1`?

Reply:
No reason, except that this test file was edited by two or three different people. This is a test code and this is sort of nit-picking, so allow me to dismiss this.

Reply:
Alright, since I reformatted the entire file I fixed this one too, so we always use `< 0` for consistency now.